utils/polus-bfio-util/bfio/backends.py

```diff
@@ -0,0 +1,787 @@
+        height = metadata.image().Pixels.get_SizeY()
+        
+        for tag in self._rdr.pages[0].tags:
+            logger.debug(tag)
+        
+        tile_size = None
+        
+        if not self._rdr.pages[0].is_tiled:
+            if width > self.frontend._TILE_SIZE or width > self.frontend._TILE_SIZE:
+                raise TypeError(frontend._file_path.name + ' is not a tiled tiff.' +
+                                ' The python backend of the BioReader only ' +
+                                'supports OME tiled tiffs. Use the java backend ' +
+                                'to load this image.')
+            
+        elif self._rdr.pages[0].tilewidth != self.frontend._TILE_SIZE or \
+            self._rdr.pages[0].tilelength != self.frontend._TILE_SIZE:
+                
+            if (width > frontend._TILE_SIZE or height > frontend._TILE_SIZE):
+                raise ValueError('Tile width and height should be {} when '.format(self.frontend._TILE_SIZE) +
+                                 'using the python backend, but found ' +
+                                 'tilewidth={} and tilelength={}. Use the java '.format(self._rdr.pages[0].tilewidth,
+                                                                                        self._rdr.pages[0].tilelength) +
+                                 'backend to read this image.')
+        
+    def read_metadata(self):
+        self.logger.debug('read_metadata(): Reading metadata...')
+        return OMEXML(self._rdr.ome_metadata)
+    
+    def _chunk_indices(self,X,Y,Z):
+        
+        self.logger.debug('_chunk_indices(): (X,Y,Z) -> ({},{},{})'.format(X,Y,Z))
+        assert len(X) == 2
+        assert len(Y) == 2
+        assert len(Z) == 2
+        
+        offsets = []
+        bytecounts = []
+        
+        ts = self.frontend._TILE_SIZE
+        
+        x_tiles = numpy.arange(X[0]//ts,numpy.ceil(X[1]/ts),dtype=int)
+        y_tile_stride = numpy.ceil(self.frontend.x/ts).astype(int)
+        
+        self.logger.debug('_chunk_indices(): x_tiles = {}'.format(x_tiles))
+        self.logger.debug('_chunk_indices(): y_tile_stride = {}'.format(y_tile_stride))
+        
+        for z in range(Z[0],Z[1]):
+            for y in range(Y[0]//ts,int(numpy.ceil(Y[1]/ts))):
+                y_offset = int(y * y_tile_stride)
+                ind = (x_tiles + y_offset).tolist()
+                
+                o = [self._rdr.pages[z].dataoffsets[i] for i in ind]
+                b = [self._rdr.pages[z].databytecounts[i] for i in ind]
+                
+                self.logger.debug('_chunk_indices(): offsets = {}'.format(o))
+                self.logger.debug('_chunk_indices(): bytecounts = {}'.format(b))
+                
+                offsets.extend(o)
+                bytecounts.extend(b)
+        
+        return offsets,bytecounts
+    
+    def _process_chunk(self, args):
+        
+        keyframe = self._keyframe
```

Comment:
`self._keyframe` is not defined in the constructor. It is assumed here that the variable exists (set in `_read_image`). I also realize this is a "private" member. However, I think it is best to define `self._keyframe = None` in the constructor and have some sort of assert in this function for readability and error handling.

Reply:
Same comment applies to other member variables below.